collector/pg_stat_statements.go

```diff
@@ -71,47 +71,43 @@ var (
 		prometheus.Labels{},
 	)
 
-	pgStatStatementsQuery = `SELECT
-		pg_get_userbyid(userid) as user,
-		pg_database.datname,
-		pg_stat_statements.queryid,
-		pg_stat_statements.calls as calls_total,
-		pg_stat_statements.total_time / 1000.0 as seconds_total,
-		pg_stat_statements.rows as rows_total,
-		pg_stat_statements.blk_read_time / 1000.0 as block_read_seconds_total,
-		pg_stat_statements.blk_write_time / 1000.0 as block_write_seconds_total
-		FROM pg_stat_statements
-	JOIN pg_database
-		ON pg_database.oid = pg_stat_statements.dbid
-	WHERE
-		total_time > (
-		SELECT percentile_cont(0.1)
-			WITHIN GROUP (ORDER BY total_time)
-			FROM pg_stat_statements
+	pgStatStatementsQuery = `WITH percentiles AS (
+    		SELECT percentile_cont(0.1) WITHIN GROUP (ORDER BY total_time) AS percentile
+    		FROM pg_stat_statements
 		)
-	ORDER BY seconds_total DESC
-	LIMIT 100;`
-
-	pgStatStatementsNewQuery = `SELECT
-		pg_get_userbyid(userid) as user,
-		pg_database.datname,
-		pg_stat_statements.queryid,
-		pg_stat_statements.calls as calls_total,
-		pg_stat_statements.total_exec_time / 1000.0 as seconds_total,
-		pg_stat_statements.rows as rows_total,
-		pg_stat_statements.blk_read_time / 1000.0 as block_read_seconds_total,
-		pg_stat_statements.blk_write_time / 1000.0 as block_write_seconds_total
-		FROM pg_stat_statements
-	JOIN pg_database
-		ON pg_database.oid = pg_stat_statements.dbid
-	WHERE
-		total_exec_time > (
-		SELECT percentile_cont(0.1)
-			WITHIN GROUP (ORDER BY total_exec_time)
-			FROM pg_stat_statements
-		)
-	ORDER BY seconds_total DESC
-	LIMIT 100;`
+		SELECT DISTINCT ON (pss.queryid, pg_get_userbyid(pss.userid), pg_database.datname)
+    		pg_get_userbyid(pss.userid) as user,
+    		pg_database.datname,
+    		pss.queryid,
+    		pss.calls as calls_total,
+    		pss.total_time / 1000.0 as seconds_total,
+    		pss.rows as rows_total,
+    		pss.blk_read_time / 1000.0 as block_read_seconds_total,
+    		pss.blk_write_time / 1000.0 as block_write_seconds_total
+		FROM pg_stat_statements pss
+		JOIN pg_database ON pg_database.oid = pss.dbid
+			CROSS JOIN percentiles
```

Comment:
Can you explain the CROSS JOIN here? I'm having trouble understanding the why. My understanding is that CROSS JOIN will produce a cartesian product where every possible combination of table A and B will exist.